alibi_detect/cd/keops/mmd.py

```diff
@@ -0,0 +1,179 @@
+            x_ref_preprocessed=x_ref_preprocessed,
+            preprocess_at_init=preprocess_at_init,
+            update_x_ref=update_x_ref,
+            preprocess_fn=preprocess_fn,
+            sigma=sigma,
+            configure_kernel_from_x_ref=configure_kernel_from_x_ref,
+            n_permutations=n_permutations,
+            input_shape=input_shape,
+            data_type=data_type
+        )
+        self.meta.update({'backend': 'keops'})
+
+        # set device
+        self.device = get_device(device)
+
+        # initialize kernel
+        sigma = torch.from_numpy(sigma).to(self.device) if isinstance(sigma,  # type: ignore[assignment]
+                                                                      np.ndarray) else None
+        self.kernel = kernel(sigma).to(self.device) if kernel == GaussianRBF else kernel
+
+        # set the correct MMD^2 function based on the batch size for the permutations
+        self.batch_size = batch_size_permutations
+        self.n_batches = 1 + (n_permutations - 1) // batch_size_permutations
+
+        # infer the kernel bandwidth from the reference data
+        if self.infer_sigma or isinstance(sigma, torch.Tensor):
+            x = torch.from_numpy(self.x_ref).to(self.device)
+            _ = self.kernel(LazyTensor(x[:, None, :]), LazyTensor(x[None, :, :]), infer_sigma=self.infer_sigma)
+            self.infer_sigma = False
+        else:
+            self.infer_sigma = True
+
+    def _mmd2(self, x_all: torch.Tensor, perms: List[torch.Tensor], m: int, n: int) \
+            -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Batched (across the permutations) MMD^2 computation for the original test statistic and the permutations.
+
+        Parameters
+        ----------
+        x_all
+            Concatenated reference and test instances.
+        perms
+            List with permutation vectors.
+        m
+            Number of reference instances.
+        n
+            Number of test instances.
+
+        Returns
+        -------
+        MMD^2 statistic for the original and permuted reference and test sets.
+        """
+        k_xx, k_yy, k_xy = [], [], []
+        for batch in range(self.n_batches):
+            i, j = batch * self.batch_size, (batch + 1) * self.batch_size
+            # construct stacked tensors with a batch of permutations for the reference set x and test set y
+            x = torch.cat([x_all[perm[:m]][None, :, :] for perm in perms[i:j]], 0)
+            y = torch.cat([x_all[perm[m:]][None, :, :] for perm in perms[i:j]], 0)
+            if batch == 0:
+                x = torch.cat([x_all[None, :m, :], x], 0)
+                y = torch.cat([x_all[None, m:, :], y], 0)
+            x, y = x.to(self.device), y.to(self.device)
+
+            # batch-wise kernel matrix computation over the permutations
+            k_xx.append(self.kernel(
+                LazyTensor(x[:, :, None, :]), LazyTensor(x[:, None, :, :])).sum(1).sum(1).squeeze(-1))
```

Comment:
Is there a reason we don't use the `LazyTensor` within the kernel? That way we'd keep the `KeOps` dependency isolated to the kernel code.

Reply:
We could but would have to add an additional kwarg to the forward pass that all keops kernels (incl. user defined ones) need to comply with to specify what type of dimension expanding for broadcasting needs to be done: `[n_x, n_y, feature_dim]` or `[batch_dim, n_x, n_y, feature_dim]`. Keeping this out of the kernel avoids this piece of custom logic. I don't feel very strongly about this (had it inside the kernel initially actually) but tried to avoid putting logic which has little to do with kernel matrix computation in the kernel forward pass. Further thoughts @jklaise @ascillitoe ?

Reply:
ok, the reason I ask is just that I'm thinking about the implementation w.r.t. the outlier detectors. It probably doesn't make sense to consider it within this PR. I'm writing up my notes on this [here](https://www.notion.so/Keops-b57624f16c01416b9a1589031202079d).

Reply:
A bit late to this one, but if it were the case that `MMDDriftKeops` and `MMDDriftTorch`detectors were almost identical, then I think it might be worth looking into isolating the KeOps functionality to the kernel (so that the KeOps backend would essentially be the torch backend with a special keops kernel). However, since`MMDDriftKeops` has many other differences (batching etc), I'm not quite sure what benefit isolating keops to the kernel code would bring? It doesn't save us much wrt to optional deps management does it? Since we're already inside `keops/` anyway... 

Deffo worth continuing the discussion wrt to outliers over on your notes!

Reply:
It's a bit different as @mauicv for outlier detectors since they can be treated as more vanilla models and don't have to deal with batching across permutations etc.

Reply:
Yep makes sense 👍🏻